lib/stdlib/test/erl_eval_SUITE.erl

```diff
@@ -1172,13 +1173,143 @@ simple1() ->
+    backtrace_check("[X || <<X>> <= 1].", {bad_generator, 1},
+                    [erl_eval, mystack(1)], none, EFH),
+    backtrace_check("<<X || X <- 1>>.", {bad_generator, 1},
+                    [erl_eval, mystack(1)], none, EFH),
+    backtrace_check("<<X || <<X>> <= 1>>.", {bad_generator, 1},
+                    [erl_eval, mystack(1)], none, EFH),
+
+    backtrace_check("if false -> true end.", if_clause,
+                    [erl_eval, mystack(1)], none, EFH),
+    backtrace_check("case 0 of 1 -> true end.", {case_clause, 0},
+                    [erl_eval, mystack(1)], none, EFH),
+    backtrace_check("try 0 of 1 -> true after ok end.", {try_clause, 0},
+                    [mystack(1)], none, EFH),
+
+    backtrace_check("fun(0) -> 1 end(1).", function_clause,
+                    [{erl_eval,'-inside-an-interpreted-fun-', [1]}, erl_eval, mystack(1)],
+                    none, EFH),
+    backtrace_check("fun F(0) -> 1 end(1).", function_clause,
+                    [{erl_eval,'-inside-an-interpreted-fun-', [1]}, erl_eval, mystack(1)],
+                    none, EFH),
+
+    fun() ->
+            {'EXIT', {{illegal_pattern,_}, BT}} =
+                (catch parse_and_run("make_ref() = 1.", none, EFH)),
+            check_backtrace([erl_eval, mystack(1)], BT)
+    end(),
+
+    %% eval_bits
+    backtrace_check("<<100:8/bitstring>>.",
+                    badarg,
+                    [{eval_bits,eval_exp_field1,8}, mystack(1)],
+                    none, EFH),
+    backtrace_check("<<100:8/foo>>.",
+                    {undefined_bittype,foo},
+                    [{eval_bits,make_bit_type,4}, mystack(1)],
+                    none, EFH),
+    backtrace_check("B = <<\"foo\">>, <<B/binary-unit:7>>.",
+                    badarg,
+                    [{eval_bits,eval_exp_field1,8}, mystack(1)],
+                    none, EFH),
+
+    ok.
+
+mystack(Line) ->
+    {my_module, my_function, 0, [{file, "evaluator"}, {line, Line}]}.
+
+custom_stacktrace_eval_handler(Ann, FunOrModFun, Args) ->
+    try
+        case FunOrModFun of
+            {Mod, Fun} -> apply(Mod, Fun, Args);
+            Fun -> apply(Fun, Args)
+        end
+    catch
+        Kind:Reason:Stacktrace ->
+            %% Take everything up to the evaluation function
+            Pruned =
+                lists:takewhile(fun
+                  ({erl_eval_SUITE,backtrace_check,5,_}) -> false;
+                  (_) -> true
+                end, Stacktrace),
+            %% Now we prune any shared code path from erl_eval
+            {current_stacktrace, Current} =
+                erlang:process_info(self(), current_stacktrace),
+            Reversed = drop_common(lists:reverse(Current), lists:reverse(Pruned)),
+            Location = [{file, "evaluator"}, {line, erl_anno:line(Ann)}],
+            %% Add our file+line information at the bottom
+            Custom = lists:reverse([{my_module, my_function, 0, Location} | Reversed]),
+            erlang:raise(Kind, Reason, Custom)
```

Comment:
For the tests, I have tried to emulate what I could do to get more precise stacktraces on eval.